text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,639 @@
+from the user.
+
+This OTEP aims at proposing consistent guidelines for creating spans that model
+the stages of the messaging flow, and correlating those in a way so that the
+requirements sketched above can be met in a consistent way across messaging
+scenarios and different messaging systems.
+
+## Terminology
+
+For terms used in this document, refer to [OTEP 173](#0173-messaging-semantic-conventions.md#terminology).
+
+## Stages of producing and consuming messages
+
+Producing and consuming a message involves five stages:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  CR[Create] --> PU[Publish]
+  end
+  subgraph INTERMEDIARY
+  direction TB
+  end
+  subgraph CONSUMER
+  direction TB
+  RE[Receive] --> PR[Process]
+  PR --> SE[Settle]
+  end
+  PU --> INTERMEDIARY
+  INTERMEDIARY --> RE
+  SE -..-> INTERMEDIARY
+```
+
+1. The producer creates a message.
+2. The producer publishes the message to an intermediary.
+3. The consumer receives the message from an intermediary.
+4. The consumer processes the message.
+5. The consumer settles the message by notifying the intermediary that the
+   message was processed. In some cases (fire-and-forget scenarios, or when
+   settlement happens on the broker), the settlement stage does not exist.
+
+The semantic conventions described below define how to model those stages with
+spans.
+
+## Trace structure
+
+### Producer
+
+Producers are responsible for injecting a creation context into a message.
+Subsequent consumers will use this context to link consumer traces to producer
+traces. Ideally, each message gets a unique and distinct creation context
+assigned. However, as a context must refer to a span this would require the
+creation of a distinct span for each message, which is not feasible in all
+scenarios. In certain batching scenarios where many messages are created and
+published in large batches, creating a span for every single message would
+obfuscate traces and is not desirable. Thus instrumentation libraries and
+auto-instrumentation should default to creating a unique and distinct context
+per message, but may support configuration or other ways to change this default
+behavior.
+
+For each producer scenario, a "Publish" span needs to be created. This span
+measures the duration of the call or operation that provides messages for
+sending or publishing to an intermediary. This call or operation (and the
+related "Publish" span) can either refer to a single message or a batch of
+multiple messages.
+
+It is recommended to create a "Create" span for every single message. "Create"
+spans can be created during the "Publish" operation as children of the
+"Publish" span. Alternatively, "Create" spans can be created independently of
+the "Publish" operation. In this case, the "Publish" span should link to the
+"Create" spans.
+
+If a "Create" span exists for a message, its context must be injected into the
```

Comment:
I'm reading this statement again and I have a few concerns..

> If a "Create" span exists for a message, its context must be injected into the message

If there's a create span, it should already be inside the message, no? 

> if no "Create" span exists for a message, the context of the related "Publish" span must be injected into the message.

This is conflicting with the statement above:

> "Create" spans can be created during the "Publish" operation as children of the "Publish" span

I think then it's unclear what should I do if messages arrive at "publish" without a context

Should instrumentation create a "create" span per message and inject that? OR should instrumentation attach the "publish" span to all messages?

Reply:
Here's one suggestion I thought of:

```
It is recommended to create a "Create" span for every single message. 
"Create" spans can be created during the "Publish" operation as children of the "Publish" span. 
Alternatively, "Create" spans can be created independently of the "Publish" operation 
(e.g., in some other parts of the application, before calling "publish"). 
In this case, the "Publish" span should link to the "Create" spans already present in the message.

Alternatively, if no "Create" span exists for a message, and during publish it is not desired
to create individual spans for each message, 
the context of the related "Publish" span must be injected into the message.
```

Reply:
I will think a bit more about how to polish this section.

My intention was to have separate paragraphs for (1) defining which spans to create, (2) defining which context to attach. I didn't want to mix up those two to avoid confusion.

Regarding context propagation, the two possibilities are:
1. A "Create" span is created for a message, either as child of "Publish" or independently of it. In any case, its context should be associated with the message.
2. No "Create" span is created. In that case, the context of the "Publish" span should be associated with the message, as this span is required to exist.

> It is recommended to create a "Create" span for every single message. 
"Create" spans can be created during the "Publish" operation as children of the "Publish" span. 
Alternatively, "Create" spans can be created independently of the "Publish" operation 
(e.g., in some other parts of the application, before calling "publish"). 
In this case, the "Publish" span should link to the "Create" spans already present in the message.
>
> Alternatively, if no "Create" span exists for a message, and during publish it is not desired
to create individual spans for each message, 
the context of the related "Publish" span must be injected into the message.

This is a bit confusing to me, as it doesn't explicitly specify what context to attach if a "Create" span is present.